vms/proposervm/proposer/windower.go

```diff
@@ -148,3 +168,125 @@ func (w *windower) Delay(ctx context.Context, chainHeight, pChainHeight uint64,
 	}
 	return delay, nil
 }
+
+func (w *windower) ExpectedProposer(
+	ctx context.Context,
+	blockHeight,
+	pChainHeight,
+	slot uint64,
+) (ids.NodeID, error) {
+	validators, err := w.sortedValidators(ctx, pChainHeight)
+	if err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	var (
+		seed             = postDurangoSeed(w.chainSource, blockHeight, slot)
+		source           = prng.NewMT19937_64()
+		validatorWeights = validatorsToWeight(validators)
+	)
+
+	sampler := sampler.NewDeterministicWeightedWithoutReplacement(source)
+	if err := sampler.Initialize(validatorWeights); err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	source.Seed(seed)
+	indices, err := sampler.Sample(1)
+	if err != nil {
+		return ids.EmptyNodeID, fmt.Errorf("%w, %w", err, ErrNoProposersAvailable)
+	}
+	return validators[indices[0]].id, nil
```

Comment:
Are we guaranteed to get at least 1 response if nil error?

Reply:
The sampler will error if the length of the slice isn't as requested (which is the provided param to sample). So yes